File: pkg/datasource/kubectl.go
package datasource

import (
	"context"
	"os/exec"
	"strings"
	"time"
)

type KubectlDataSource struct {
	callbacks     []func(interface{})
	currentValue  interface{}
	updateChannel chan KubectlData
}

func NewKubectlDataSource() *KubectlDataSource {
	return &KubectlDataSource{
		callbacks:     make([]func(interface{}), 0),
		updateChannel: make(chan KubectlData, 1),
	}
}

func (k *KubectlDataSource) Start(ctx context.Context) {
	go func() {
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()

		// Initial fetch
		k.updateContext()

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				k.updateContext()
			case data := <-k.updateChannel:
				k.currentValue = data
				for _, callback := range k.callbacks {
					callback(data)
				}
			}
		}
	}()
}

func (k *KubectlDataSource) Subscribe(callback func(value interface{})) {
	k.callbacks = append(k.callbacks, callback)
}

func (k *KubectlDataSource) GetCurrentValue() interface{} {
	return k.currentValue
}

func (k *KubectlDataSource) ClearContext() {
	// Clear the current context by unsetting it
	cmd := exec.Command("kubectl", "config", "unset", "current-context")
	cmd.Run() // Ignore errors since context might already be unset
	
	// Immediately update to reflect the change
	k.updateContext()
}

func (k *KubectlDataSource) GetAllContexts() []string {
	cmd := exec.Command("kubectl", "config", "get-contexts", "-o", "name")
	output, err := cmd.Output()
	if err != nil {
		return []string{}
	}

	contexts := strings.Split(strings.TrimSpace(string(output)), "\n")
	var validContexts []string
	for _, context := range contexts {
		context = strings.TrimSpace(context)
		if context != "" {
			validContexts = append(validContexts, context)
		}
	}
	
	return validContexts
}

func (k *KubectlDataSource) SetContext(context string) {
	cmd := exec.Command("kubectl", "config", "use-context", context)
	cmd.Run() // Ignore errors
	
	// Immediately update to reflect the change
	k.updateContext()
}

func (k *KubectlDataSource) updateContext() {
	context := getKubectlContext()
	isProduction := isProductionContext(context)
	data := KubectlData{
		Context:       context,
		ContextString: "⚙️ " + context,
		IsProduction:  isProduction,
	}

	select {
	case k.updateChannel <- data:
	default:
		// Channel full, skip update
	}
}

func getKubectlContext() string {
	cmd := exec.Command("kubectl", "config", "current-context")
	output, err := cmd.Output()
	if err != nil {
		return "--"
	}

	context := strings.TrimSpace(string(output))
	if context == "" {
		return "--"
	}

	return context
}

func isProductionContext(context string) bool {
	contextLower := strings.ToLower(context)
	return strings.Contains(contextLower, "prod") || strings.Contains(contextLower, "prd")
}

File: src/sources/kubectl.rs
//! Kubectl current-context. Port of pkg/datasource/kubectl.go.

use std::process::Command;

#[derive(Debug, Clone)]
pub struct KubectlData {
    pub context: String,
    pub string: String,
    pub is_production: bool,
}

impl Default for KubectlData {
    fn default() -> Self {
        KubectlData {
            context: "--".to_string(),
            string: "⚙️ --".to_string(),
            is_production: false,
        }
    }
}

pub fn update_context() -> KubectlData {
    let context = get_kubectl_context();
    let is_production = is_production_context(&context);
    KubectlData {
        string: format!("⚙️ {}", context),
        is_production,
        context,
    }
}

pub fn get_kubectl_context() -> String {
    let output = Command::new("kubectl")
        .args(["config", "current-context"])
        .output();
    match output {
        Ok(o) if o.status.success() => {
            let ctx = String::from_utf8_lossy(&o.stdout).trim().to_string();
            if ctx.is_empty() {
                "--".to_string()
            } else {
                ctx
            }
        }
        _ => "--".to_string(),
    }
}

pub fn is_production_context(context: &str) -> bool {
    let lower = context.to_lowercase();
    lower.contains("prod") || lower.contains("prd")
}

pub fn get_all_contexts() -> Vec<String> {
    let output = Command::new("kubectl")
        .args(["config", "get-contexts", "-o", "name"])
        .output();
    match output {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout)
            .trim()
            .lines()
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect(),
        _ => Vec::new(),
    }
}

pub fn set_context(context: &str) {
    let _ = Command::new("kubectl")
        .args(["config", "use-context", context])
        .status();
}

pub fn clear_context() {
    let _ = Command::new("kubectl")
        .args(["config", "unset", "current-context"])
        .status();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn production_detection() {
        assert!(is_production_context("prod-cluster"));
        assert!(is_production_context("my-PROD-eu"));
        assert!(is_production_context("acme-prd-1"));
        assert!(!is_production_context("staging"));
        assert!(!is_production_context("dev-local"));
        assert!(!is_production_context("--"));
    }
}
